tower/src/util/rng.rs

```diff
@@ -0,0 +1,198 @@
+//! [PRNG] utilities for tower middleware.
+//!
+//! This module provides a generic [`Rng`] trait and a [`HasherRng`] that
+//! implements the trait based on [`RandomState`] or any other [`Hasher`].
+//!
+//! These utlities replace tower's internal usage of `rand` with these smaller
+//! more light weight methods. Most of the implemenations are extracted from
+//! their corresponding `rand` implementations.
+//!
+//! [PRNG]: https://en.wikipedia.org/wiki/Pseudorandom_number_generator
+
+use std::{
+    collections::hash_map::RandomState,
+    hash::{BuildHasher, Hasher},
+    ops::Range,
+};
+
+/// A simple [PRNG] trait for use within tower middleware.
+///
+/// [PRNG]: https://en.wikipedia.org/wiki/Pseudorandom_number_generator
+pub trait Rng {
+    /// Generate a random [`u64`].
+    fn next_u64(&mut self) -> u64;
+
+    /// Generate a random [`f64`] between `[0, 1)`.
+    fn next_f64(&mut self) -> f64 {
+        // Borrowed from:
+        // https://github.com/rust-random/rand/blob/master/src/distributions/float.rs#L106
+        let float_size = std::mem::size_of::<f64>() as u32 * 8;
+        let precison = 52 + 1;
+        let scale = 1.0 / ((1u64 << precison) as f64);
+
+        let value = self.next_u64();
+        let value = value >> (float_size - precison);
+
+        scale * value as f64
+    }
+
+    /// Randomly pick a value within the range.
+    ///
+    /// # Panic
+    ///
+    /// - If start < end this will panic in debug mode.
+    fn next_range(&mut self, range: Range<u64>) -> u64 {
+        debug_assert!(
+            range.start < range.end,
+            "The range start must be smaller than the end"
+        );
+        let start = range.start;
+        let end = range.end;
+
+        let range = end - start;
+
+        let n = self.next_u64();
+
+        (n % range) + start
+    }
+}
+
+impl<R: Rng + ?Sized> Rng for Box<R> {
+    fn next_u64(&mut self) -> u64 {
+        (**self).next_u64()
+    }
+}
+
+/// A [`Rng`] implementation that uses a [`Hasher`] to generate the random
+/// values. The implementation uses an internal counter to pass to the hasher
+/// for each iteration of [`Rng::next_u64`].
+///
+/// # Default
+///
+/// This hasher has a default type of [`RandomState`] which just uses the
+/// libstd method of getting a random u64.
+#[derive(Debug)]
+pub struct HasherRng<H = RandomState> {
+    hasher: H,
+    counter: u64,
+}
+
+impl HasherRng {
+    /// Create a new default [`HasherRng`].
+    pub fn new() -> Self {
+        HasherRng::default()
+    }
+}
+
+impl Default for HasherRng {
+    fn default() -> Self {
+        HasherRng::with_hasher(RandomState::default())
+    }
+}
+
+impl<H> HasherRng<H> {
+    /// Create a new [`HasherRng`] with the provided hasher.
+    pub fn with_hasher(hasher: H) -> Self {
+        HasherRng { hasher, counter: 0 }
+    }
+}
+
+impl<H> Rng for HasherRng<H>
+where
+    H: BuildHasher,
+{
+    fn next_u64(&mut self) -> u64 {
+        let mut hasher = self.hasher.build_hasher();
```

Comment:
I don't think you want to `build_hasher` for each number. That will restart the hasher back to the beginning. I'd expect you'd build a hasher once, and then feed the counter into it over and over. Or am I wrong?

Reply:
I thought finished consumed it, but I guess finish just takes a &self. Let me see if that makes any meaningful difference.

Reply:
Yea, thinking more, I could be much more wrong than right.

Reply:
Follow up here HashMap creates a hasher for each hash so we are reverting back to that setup.